flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java

```diff
@@ -0,0 +1,263 @@
+            long periodicMaterializeInitDelay,
+            long periodicMaterializeInterval,
+            int allowedNumberOfFailures,
+            boolean materializationEnabled) {
+        this.mailboxExecutor = mailboxExecutor;
+        this.asyncOperationsThreadPool = asyncOperationsThreadPool;
+        this.keyedStateBackend = keyedStateBackend;
+        this.stateChangelogWriter = stateChangelogWriter;
+        this.asyncExceptionHandler = asyncExceptionHandler;
+        this.periodicExecutor =
+                Executors.newSingleThreadScheduledExecutor(
+                        new ExecutorThreadFactory("periodic-materialization"));
+        this.streamFactory = shared -> checkpointStorageWorkerView.createTaskOwnedStateStream();
+        this.allowedNumberOfFailures = allowedNumberOfFailures;
+        this.materializationOnGoing = new AtomicBoolean(false);
+        this.retries = new AtomicInteger(allowedNumberOfFailures);
+
+        this.materializedId = 0;
+        this.materializedState = materializedState;
+
+        if (materializationEnabled) {
+            this.periodicExecutor.scheduleAtFixedRate(
+                    this::triggerMaterialization,
+                    periodicMaterializeInitDelay,
+                    periodicMaterializeInterval,
+                    TimeUnit.MILLISECONDS);
+        }
+    }
+
+    @VisibleForTesting
+    public void triggerMaterialization() {
+        mailboxExecutor.execute(
+                () -> {
+                    // Only one materialization ongoing at a time
+                    if (!materializationOnGoing.compareAndSet(false, true)) {
+                        return;
+                    }
+
+                    SequenceNumber upTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
+
+                    if (upTo.equals(materializedState.lastMaterializedTo())) {
+                        return;
+                    }
+
+                    RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture =
+                            keyedStateBackend.snapshot(
+                                    materializedId++, // TO DISCUSS: seems this does not matter,
+                                    // based on
+                                    System.currentTimeMillis(),
+                                    streamFactory,
+                                    CHECKPOINT_OPTIONS);
+
+                    // TODO: add metadata to log FLINK-23170
+                    asyncOperationsThreadPool.execute(
+                            () -> asyncMaterialization(materializedRunnableFuture, upTo));
+                },
+                "materialization");
+    }
+
+    private void asyncMaterialization(
+            RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture,
+            SequenceNumber upTo) {
+
+        FileSystemSafetyNet.initializeSafetyNetForThread();
+        try {
+            FutureUtils.runIfNotDoneAndGet(materializedRunnableFuture);
```

Comment:
Can we close the safety net right after this statement (`runIfNotDoneAndGet`)?
I think it would make the flow easier to follow and possible exceptions won't affect each other.

Reply:
but when exceptions happen, we also need to close the safety net, that's why put within `finally`

Reply:
`mailboxExecutor.execute` and everything that's inside doesn't need this.
Then maybe split `asyncMaterialization` function into two:
  - `materialize` (try/finally with safety net)
  - `mailboxExecutor.execute` (and if result is `null`/empty then only update `materializationOnGoing` - in the same mail)
  ?